### DISCUSSION TITLE
logrotate (USR1 signal) restarts uvicorn worker

### DISCUSSION THREAD

**Discussion by wonjoonSeol-WS** [Kludex/uvicorn#1558] [Potential Issue] [answered]
I use FastAPI+Uvicorn+Gunicorn and issuing USR1 signal to uvicorn worker restarts the worker.

I believe the issue is to do with uvicorn worker not implementing signal handler - [github.com/encode/uvicorn/blob/…](https://github.com/encode/uvicorn/blob/291c195c65940d33025d91778758f06924f34bfd/uvicorn/workers.py#L68). 

```
    def init_signals(self) -> None:
        # Reset signals so Gunicorn doesn't swallow subprocess return codes
        # other signals are set up by Server.install_signal_handlers()
        # See: https://github.com/encode/uvicorn/issues/894
        for s in self.SIGNALS:
            signal.signal(s, signal.SIG_DFL)
```

It converts it to signal.SIG_DFL. 

SIGUSR1 and SIGUSR2 both have the default action Term -- the process is terminated. ([unix.stackexchange.com/questions/38589/…](https://unix.stackexchange.com/questions/38589/why-does-sigusr1-cause-process-to-be-terminated)) 

According to a similar issue page, https://github.com/encode/uvicorn/issues/896#issuecomment-762624434
[yinkh](https://github.com/yinkh) changed init_signals to gunicorn worker's original and this change no longer restarts the worker.

I also notice that Uvicorn only handles two signals:
```
HANDLED_SIGNALS = (
    signal.SIGINT,  # Unix signal 2. Sent by Ctrl+C.
    signal.SIGTERM,  # Unix signal 15. Sent by `kill <pid>`.
)
```

I would be more than happy to contribute to uvicorn, if I could change the worker init_signals function to do the same with gunicorn worker?

```
    def init_signals(self):
        # reset signaling
        for s in self.SIGNALS:
            signal.signal(s, signal.SIG_DFL)
        # init new signaling
        signal.signal(signal.SIGQUIT, self.handle_quit)
        signal.signal(signal.SIGTERM, self.handle_exit)
        signal.signal(signal.SIGINT, self.handle_quit)
        signal.signal(signal.SIGWINCH, self.handle_winch)
        signal.signal(signal.SIGUSR1, self.handle_usr1)
        signal.signal(signal.SIGABRT, self.handle_abort)

        # Don't let SIGTERM and SIGUSR1 disturb active requests
        # by interrupting system calls
        signal.siginterrupt(signal.SIGTERM, False)
        signal.siginterrupt(signal.SIGUSR1, False)

        if hasattr(signal, 'set_wakeup_fd'):
            signal.set_wakeup_fd(self.PIPE[1])
```

**Comment by euri10** [comment, marked as answer]
seems legit issue to me, coming from https://stackoverflow.com/questions/72891901/gunicorn-worker-terminated-due-to-signal-10
I guess the https://github.com/encode/uvicorn/pull/895 was a little bit too much on resetting signal for that usecase
PR welcomed

**Comment by wonjoonSeol-WS** [comment]
Added steps to reproduce and a potential solution here:
https://github.com/encode/uvicorn/issues/1559#issuecomment-1179776669
